regcore/db/django_models.py

```diff
@@ -1,46 +1,86 @@
 """Each of the data structures relevant to the API (regulations, notices,
 etc.), implemented using Django models"""
+import collections
+
 from django.core.exceptions import ObjectDoesNotExist
 
 from regcore.models import Diff, Layer, Notice, Regulation
 
 
+def get_tree(regs):
+    tree = collections.defaultdict(list)
+    for reg in regs:
+        if reg.parent_id is not None:
+            tree[reg.parent_id].append(reg)
+    return tree
+
+
 class DMRegulations(object):
     """Implementation of Django-models as regulations backend"""
     def get(self, label, version):
         """Find the regulation label + version"""
-        try:
-            reg = Regulation.objects.get(version=version,
-                                         label_string=label)
-            as_dict = {
-                'label': reg.label_string.split('-'),
-                'text': reg.text,
-                'node_type': reg.node_type,
-                'children': reg.children
-            }
-            if reg.title:
-                as_dict['title'] = reg.title
-            return as_dict
-        except ObjectDoesNotExist:
+        regs = Regulation.objects.filter(
+            version=version,
+            label_string=label,
+        ).get_descendants(
+            include_self=True,
+        )
+        regs = list(regs.all())
+        if not regs:
             return None
-
-    def _transform(self, reg, version):
+        tree = get_tree(regs)
+        return self.serialize(regs[0], tree)
+
+    def serialize(self, reg, tree):
+        ret = {
+            'label': reg.label_string.split('-'),
+            'text': reg.text,
+            'node_type': reg.node_type,
+            'children': [
+                self.serialize(child, tree)
+                for child in tree.get(reg.id, [])
+            ],
+        }
+        if reg.title:
+            ret['title'] = reg.title
+        return ret
+
+    def _transform(self, reg, version, parent=None):
         """Create the Django object"""
-        return Regulation(version=version,
-                          label_string='-'.join(reg['label']),
-                          text=reg['text'],
-                          title=reg.get('title', ''),
-                          node_type=reg['node_type'],
-                          root=(len(reg['label']) == 1),
-                          children=reg['children'])
+        return Regulation(
+            id=self._get_id(reg, version),
+            parent_id=self._get_id(parent, version) if parent else None,
+            version=version,
+            label_string='-'.join(reg['label']),
+            text=reg['text'],
+            title=reg.get('title', ''),
+            node_type=reg['node_type'],
+            root=(len(reg['label']) == 1),
+        )
+
+    def _get_id(self, reg, version):
+        return '{}-{}'.format(version, '-'.join(reg['label']))
 
     def bulk_put(self, regs, version, root_label):
         """Store all reg objects"""
```

Comment:
If we're changing the signature here (regs used to be a list of dicts), let's update the docstring to make it clear what's going on.